### PR TITLE
Agree to terms in the deployment flow

### PR DESCRIPTION
**Pull request by huwshimi**

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- add a charm with terms you haven't already agreed to
- open the df
- scroll down and check the box to agree to the terms
- click deploy
- the model should deploy
- add another service using the same charm
- load the deployment flow
- scroll down and the checkbox to agree should no longer appear as you've already agreed to those terms

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6692/

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 1]
```diff
@@ -69,21 +68,42 @@ YUI.add('deployment-flow', function() {
...
       if (modelName) {
         modelName.focus();
       }
     },
 
+    componentWillReceiveProps: function(nextProps) {
+      const newApps = nextProps.applications;
+      const currentApps = this.props.applications;
+      if (newApps.length !== currentApps.length ||
+        newApps.filter(a => currentApps.indexOf(a) === -1).length > 0) {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: Slick.

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 2]
```diff
@@ -69,21 +68,42 @@ YUI.add('deployment-flow', function() {
         credential: this.props.credential,
         loggedIn: !!this.props.getAuth(),
         modelName: this.props.modelName,
+        newTerms: [],
         region: this.props.region,
         showChangelogs: false,
         sshKey: null,
-        terms: this._getTerms(),
+        terms: this._getTerms() || [],
+        termsList: [],
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: We've got three different terms-ish items in state; maybe in-line comments on each to provide a little clarification on how they differ?

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 3]
```diff
@@ -299,7 +329,19 @@ YUI.add('deployment-flow', function() {
       if (this.state.sshKey) {
         args.config = {'authorized-keys': this.state.sshKey};
       }
-      this.props.deploy(this._handleClose, true, this.state.modelName, args);
+      const deploy = this.props.deploy.bind(
+        this, this._handleClose, true, this.state.modelName, args);
+      if (this.state.newTerms.length > 0) {
+        this.props.addAgreement(this.state.termsList, (error, response) => {
+          if (error) {
+            console.error('Could not agree to terms:', error);
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: Seems like we need to percolate addNotification into the component so we can display an error to the user here.

File: jujugui/static/gui/src/app/components/deployment-flow/test-deployment-flow.js [hunk 4]
```diff
@@ -1220,25 +1364,31 @@ describe('DeploymentFlow', function() {
...
       <juju.components.DeploymentFlow
         acl={acl}
+        addAgreement={sinon.stub()}
+        applications={[]}
         changes={{}}
         changesFilterByParent={sinon.stub()}
         changeState={changeState}
         charmsGetById={charmsGetById}
         deploy={deploy}
+        environment={{}}
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/deployment-flow/test-deployment-flow.js]: Thanks for the cleanup on these additional, missing props.

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js (hunk 1 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: Indeed but a comment here so futuredev doesn't have to grok the code :)

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js (hunk 3 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: `Unable to register terms agreement:  ${error}`

and then `return` here

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 5]
```diff
@@ -299,7 +329,19 @@ YUI.add('deployment-flow', function() {
...
+          } else {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: then you can remove this `else` indentation

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 6]
```diff
@@ -337,6 +380,55 @@ YUI.add('deployment-flow', function() {
     },
 
     /**
+      Get the list of terms that the user has already agreed to.
+
+      @method _getAgreements
+    */
+    _getAgreements: function() {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: I feel like this isn't being tested....did you miss a test or am I missing where it's tested?

Comment (huwshimi) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: That method is not being tested explicitly, but tests such as https://github.com/huwshimi/juju-gui/blob/c92e462e31d19abfea52c48c4a3227420f7012b2/jujugui/static/gui/src/app/components/deployment-flow/test-deployment-flow.js#L862 rely on the result of that method to pass.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6702/

**Comment by huwshimi** [comment]
@kadams54 @hatched changes made, thanks for the reviews.

**Comment by Lukewh** [comment]
QA Bad:
I charms with terms do not deploy.
Error on deploy:
```
'Could not add requested charm. Server responded with: cannot retrieve charm "cs:~cmars/trusty/canonical-terms-example-1": cannot get archive: cannot get discharge from "https://api.jujucharms.com/identity/v1/discharger": cannot start interactive session: interaction required but not possible'
10:37 AM 'Could not deploy the requested application. Server responded with: charm "cs:~cmars/trusty/canonical-terms-example-1" not found'
10:37 AM 'Could not add the requested unit. Server responded with: application "canonical-terms-example" not found'
```

**Comment by hatched** [comment]
I investigated this further and it appears that the original /agreement POST requires a macaroon to be discharged. The Bakery then goes and does its dance, however when it tries to re-make the call it makes it as a GET not a POST.

https://github.com/juju/juju-gui/blob/develop/jujugui/static/gui/src/app/store/env/bakery.js#L421

So this method needs to be upgraded to take the original request type and body and make the appropriate call.

**Comment by hatched** [comment]
QA Issue:

Also I noticed that the agreement checkbox is shown before you select a cloud - it should have the same visibility modifiers on it as the other elements in the DF.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6723/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6748/

**Comment by hatched** [comment]
Thanks for the continued work on this @huwshimi 

When deploying a charm with 'owned terms' the `agreement` call fails with a 404.

A request is made to get the term like: https://www.jujugui.org/terms/v1/terms/hatch/test-term1
which returns:
```JavaScript
[{
  "id":"hatch/test-term1/1",
  "owner":"hatch",
  "name":"test-term1",
  "revision":1,"created-on":"2017-02-13T20:19:20Z",
  "published":true,
  "content":"I am term1\n"
}]
```
Then the request to agree body is as follows:
```JavaScript
[{
  "termname":"test-term1",
  "termrevision":1
},{
  "termname":"test-term2",
  "termrevision":1
}]
```

Because the terms have an `owner` the agreement step needs to include a `termowner` field. If there is no `owner` field then do not include the `termowner` field.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6757/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6758/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6759/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6760/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6761/

**Comment by huwshimi** [comment]
While trying to fix #2436 I'm hitting a test failure, it looks like something is being cached as it loops through the tests, will attempt another look in the morning.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6770/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6780/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6781/

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**